jenkins/spark-nightly-build.sh

```diff
@@ -89,39 +91,96 @@ function distWithReducedPom {
+    flock -x -w 300 200 || { echo "Lock acquisition failed"; exit 1; }
+
+    echo "Copying sql-plugin-api,aggregator to .m2 repo..."
+    for mod in \
+      "rapids-4-spark-sql-plugin-api_${SCALA_BINARY_VER}" "rapids-4-spark-aggregator_${SCALA_BINARY_VER}"; do
+      SRC_DIR="${SHIM_M2DIR}/com/nvidia/${mod}/${ART_VER}"
+      DEST_DIR="${M2DIR}/com/nvidia/${mod}/${ART_VER}"
+
+      if [[ -d "$SRC_DIR" ]]; then
+        mkdir -p "$DEST_DIR"
+        rsync -av --checksum --ignore-existing "${SRC_DIR}/" "${DEST_DIR}/"
+      fi
+    done
+  ) 200>"${M2DIR}/.copy.lock"
+
+  if [[ $SKIP_DEPLOY != 'true' ]]; then
+    local DEPLOY_CMD="$MVN -B deploy -pl $DEPLOY_SUBMODULES $MVN_URM_MIRROR \
+          -Dmaven.repo.local=$SHIM_M2DIR \
+          -Dcuda.version=$DEFAULT_CUDA_CLASSIFIER \
+          -DskipTests -Drat.skip=true -Djava.io.tmpdir=${SHIM_WORKSPACE} \
+          -Dmaven.scaladoc.skip -Dmaven.scalastyle.skip=true \
+          -Dbuildver=${BUILD_VER}"
+
+    echo "Deploying ${BUILD_VER}"
+    DEPLOY_LOG="${SHIM_WORKSPACE}/deploy-${BUILD_VER}.log"
+
+    $DEPLOY_CMD > "${DEPLOY_LOG}" 2>&1
+    CODE="$?"
+    if [[ $CODE != "0" ]]; then
+      tail -1000 "${DEPLOY_LOG}" || true
+      return $CODE
+    fi
+  fi
+
+  echo "${BUILD_VER} done."
+}
+export -f build_shim
+
 # build, install, and deploy all the versions we support, but skip deploy of individual dist module since we
 # only want the combined jar to be pushed.
 # Note this does not run any integration tests
 # Deploy jars unless SKIP_DEPLOY is 'true'
 
-# option to skip unit tests. Used in our CI to separate test runs in parallel stages
-SKIP_TESTS=${SKIP_TESTS:-"false"}
-
 set +H # turn off history expansion
-DEPLOY_SUBMODULES=${DEPLOY_SUBMODULES:-"integration_tests"}
-for buildver in "${SPARK_SHIM_VERSIONS[@]:1}"; do
-    $MVN -U -B clean install $MVN_URM_MIRROR -Dmaven.repo.local=$M2DIR \
-        -Dcuda.version=$DEFAULT_CUDA_CLASSIFIER \
-        -DskipTests=$SKIP_TESTS \
-        -Dbuildver="${buildver}"
-    if [[ $SKIP_TESTS == "false" ]]; then
-      # Run filecache tests
-      SPARK_CONF=spark.rapids.filecache.enabled=true \
-          $MVN -B test -rf tests $MVN_URM_MIRROR -Dmaven.repo.local=$M2DIR \
-              -Dcuda.version=$DEFAULT_CUDA_CLASSIFIER \
-              -Dbuildver="${buildver}" \
-              -DwildcardSuites=org.apache.spark.sql.rapids.filecache.FileCacheIntegrationSuite
-    fi
-    distWithReducedPom "install"
-    [[ $SKIP_DEPLOY != 'true' ]] && \
-        # this deploys selected submodules
-        $MVN -B deploy -pl $DEPLOY_SUBMODULES $MVN_URM_MIRROR \
-            -Dmaven.repo.local=$M2DIR \
-            -Dcuda.version=$DEFAULT_CUDA_CLASSIFIER \
-            -DskipTests \
-            -Dmaven.scaladoc.skip -Dmaven.scalastyle.skip=true \
-            -Dbuildver="${buildver}"
-done
+export DEPLOY_SUBMODULES=${DEPLOY_SUBMODULES:-"integration_tests"}
+if ! command -v parallel &> /dev/null; then
```

Comment:
We may pre-install the JNI or private dep jars into the M2DIR, 

We also run this scripts in the docker container with `parallel` pre-installed, 

so better to have another condition check  like  if [ "$BUILD_PARALLELISM" le 1 ] to specify mvn build running in sequentially build.

Reply:
> We may pre-install the JNI or private dep jars into the M2DIR

Yes, this could help save some bandwidth. We can follow up later to see how much time it could potentially save

> so better to have another condition check like if [ "$BUILD_PARALLELISM" le 1 ] to specify mvn build running in sequentially build.

As this script is for jenkins only, we will not spend more time supporting sequential run IMHO. The if expression here is only to provide sanity run only